FV3/coarse_graining/coarse_graining.F90

```diff
@@ -421,6 +422,50 @@ subroutine vertically_remap_field(phalf_in, field, phalf_out, ptop, field_out)
     enddo
   end subroutine vertically_remap_field
 
+
+  subroutine vertically_remap_temperature(phalf_in, temperature, phalf_out, ptop, temperature_out)
+    real, intent(in) :: phalf_in(is:ie,js:je,1:npz+1), phalf_out(is:ie,js:je,1:npz+1)
+    real, intent(in) :: temperature(is:ie,js:je,1:npz)
+    real, intent(in) :: ptop
+    real, intent(out) :: temperature_out(is:ie,js:je,1:npz)
+
+    real, allocatable :: pfull_in(:,:,:), pfull_out(:,:,:)
+    integer :: kn, km, kord, iv, j, q2, k
+    real :: poisson_constant
+
+    poisson_constant = RDGAS / CP_AIR
+
+    kn = npz
+    km = npz
+
+    ! Hard code values of kord and iv for now
+    kord = 1
+    iv = 1
+    q2 = 1
+
+    ! Start by standardly remapping temperature using nearest-neighbor
+    ! extrapolation for all extrapolated points.
+    do j = js, je
+       call mappm(km, phalf_in(is:ie,j,:), temperature(is:ie,j,:), kn, &
+            phalf_out(is:ie,j,:), temperature_out(is:ie,j,:), is, ie, iv, kord, ptop)
+    enddo
+
+    ! Where we need to extrapolate the temperature, replace adjust the
+    ! extrapolated temperature adiabatically to the coarse level midpoint.
+    allocate(pfull_in(is:ie,js:je,1:npz))
+    allocate(pfull_out(is:ie,js:je,1:npz))
+    call compute_pfull_from_phalf(phalf_in, pfull_in)
+    call compute_pfull_from_phalf(phalf_out, pfull_out)
+
+    do k = 1, npz
+      ! Note this is the inverse condition to what is used in mask_area_weights
+      ! since we want points that *are* extrapolated rather than not.
+      where (phalf_out(is:ie,js:je,k+1) .ge. phalf_in(is:ie,js:je,npz+1))
+        temperature_out(is:ie,js:je,k) = temperature_out(is:ie,js:je,k) * (pfull_out(is:ie,js:je,k) / pfull_in(is:ie,js:je,npz)) ** poisson_constant
```

Comment:
I think it would be better to have the calculation of potential temperature in a separate subroutine like you did in fv3net. But now I'm confused as to what temperature is passed in this subroutine, is it already potential temperature?

Reply:
Sure thing, will refactor.